openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb

```diff
@@ -0,0 +1,789 @@
+    "\n",
+    "    trainer = Trainer(gpus=AVAIL_GPUS, max_epochs=1, callbacks=[MetricsCallback()])\n",
+    "\n",
+    "    trainer.validate(model=model, dataloaders=val_loader)\n",
+    "    print(\"validation logged metrics\", trainer.logged_metrics)\n",
+    "\n",
+    "    val_loss = trainer.logged_metrics[\"Discriminator val loss\"]\n",
+    "\n",
+    "    return {\"val_loss\": val_loss}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "derived-bride",
+   "metadata": {},
+   "source": [
+    "## Time to start a federated learning experiment"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "mature-renewal",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# create an experimnet in federation\n",
+    "experiment_name = \"PL_MNIST_test_experiment\"\n",
+    "fl_experiment = FLExperiment(federation=federation, experiment_name=experiment_name)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "lightweight-causing",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# The following command zips the workspace and python requirements to be transfered to collaborator nodes\n",
+    "fl_experiment.start(\n",
+    "    model_provider=MI,\n",
+    "    task_keeper=TI,\n",
+    "    data_loader=fed_dataset,\n",
+    "    rounds_to_train=5,\n",
+    "    opt_treatment=\"CONTINUE_GLOBAL\",\n",
+    "    device_assignment_policy=\"CUDA_PREFERRED\",\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f1543a36",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# If user wants to stop IPython session, then reconnect and check how experiment is going\n",
+    "# fl_experiment.restore_experiment_state(MI)\n",
+    "\n",
+    "fl_experiment.stream_metrics()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f987d2c4",
+   "metadata": {},
+   "source": [
+    "## Check the images generated by the model in tensorboard"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "fec7a708",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!pip install -r ../envoy/sd_requirements.txt\n",
+    "import sys\n",
+    "\n",
+    "sys.path.insert(1, \"../envoy\")\n",
+    "from mnist_shard_descriptor import MnistShardDescriptor"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "2bc77cd8",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "fed_dataset = MnistFedDataset(train_bs=256, valid_bs=64)\n",
+    "fed_dataset.shard_descriptor = MnistShardDescriptor(rank_worldsize=\"1,1\")\n",
+    "\n",
+    "last_model = fl_experiment.get_last_model()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "5f821972",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "imgs = fed_dataset.get_valid_data()\n",
+    "\n",
+    "val_imgs = []\n",
+    "for x in range(0, 10):\n",
```

Comment:
Range up to 10 but we see 6 images, It's a bit non-intuitive. So I guess we can remove range by this:
```python3
val_imgs = fed_dataset.get_valid_data()
```